File: server/routes/admin_routes.py
from flask import Blueprint, request, jsonify, g
from database import get_db_connection, hash_password
from middleware import token_required, admin_required

admin_bp = Blueprint("admin", __name__, url_prefix="/api/admin")

# Enforce Admin security globally across all endpoints in this blueprint
@admin_bp.before_request
@token_required
@admin_required
def enforce_admin_access():
    pass

@admin_bp.route("/dashboard-summary", methods=["GET"])
def get_dashboard_summary():
    """Retrieve top-level college metrics for Admin overview."""
    conn = get_db_connection()
    cursor = conn.cursor()

    cursor.execute("SELECT COUNT(*) FROM departments;")
    dept_count = cursor.fetchone()[0]

    cursor.execute("SELECT COUNT(*) FROM users WHERE role = 'staff';")
    staff_count = cursor.fetchone()[0]

    cursor.execute("SELECT COUNT(*) FROM students;")
    student_count = cursor.fetchone()[0]

    cursor.execute("SELECT COUNT(*) FROM marks;")
    marks_recorded = cursor.fetchone()[0]

    cursor.execute("SELECT COUNT(*) FROM notification_logs WHERE status = 'SENT';")
    notifications_sent = cursor.fetchone()[0]

    conn.close()

    return jsonify({
        "departments": dept_count,
        "staff": staff_count,
        "students": student_count,
        "marks_recorded": marks_recorded,
        "notifications_sent": notifications_sent
    }), 200

@admin_bp.route("/academic-structure", methods=["GET"])
def get_academic_structure():
    """Retrieve full hierarchy: Academic Years, Departments, Years, and Sections."""
    conn = get_db_connection()
    cursor = conn.cursor()

    cursor.execute("SELECT * FROM academic_years ORDER BY id DESC;")
    acad_years = [dict(r) for r in cursor.fetchall()]

    cursor.execute("SELECT * FROM departments ORDER BY dept_code ASC;")
    depts = [dict(r) for r in cursor.fetchall()]

    cursor.execute("SELECT * FROM years ORDER BY year_number ASC;")
    years = [dict(r) for r in cursor.fetchall()]

    cursor.execute("""
        SELECT s.id, s.section_name, s.department_id, d.dept_code, s.year_id, y.year_name, s.academic_year_id, ay.year_name as academic_year_name
        FROM sections s
        JOIN departments d ON s.department_id = d.id
        JOIN years y ON s.year_id = y.id
        JOIN academic_years ay ON s.academic_year_id = ay.id
        ORDER BY d.dept_code, y.year_number, s.section_name;
    """)
    sections = [dict(r) for r in cursor.fetchall()]

    cursor.execute("""
        SELECT sub.id, sub.subject_code, sub.subject_name, sub.department_id, d.dept_code, sub.year_id, sub.semester
        FROM subjects sub
        JOIN departments d ON sub.department_id = d.id
        ORDER BY sub.subject_code;
    """)
    subjects = [dict(r) for r in cursor.fetchall()]

    conn.close()

    return jsonify({
        "academic_years": acad_years,
        "departments": depts,
        "years": years,
        "sections": sections,
        "subjects": subjects
    }), 200

@admin_bp.route("/departments", methods=["POST"])
def create_department():
    """Register a new college department."""
    data = request.get_json() or {}
    code = data.get("dept_code", "").strip().upper()
    name = data.get("dept_name", "").strip()

    if not code or not name:
        return jsonify({"error": "Department code and name are required"}), 400

    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute("INSERT INTO departments (dept_code, dept_name) VALUES (?, ?);", (code, name))
        conn.commit()
        new_id = cursor.lastrowid
        conn.close()
        return jsonify({"message": "Department created successfully", "id": new_id, "dept_code": code}), 201
    except Exception as e:
        conn.close()
        return jsonify({"error": f"Failed to create department: {str(e)}"}), 400

@admin_bp.route("/staff", methods=["GET"])
def get_staff_list():
    """List all teaching faculty and their current allocations."""
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("""
        SELECT u.id, u.username, u.full_name, u.email, u.department_id, d.dept_name, d.dept_code, u.created_at
        FROM users u
        LEFT JOIN departments d ON u.department_id = d.id
        WHERE u.role = 'staff'
        ORDER BY u.full_name ASC;
    """)
    staff = [dict(r) for r in cursor.fetchall()]

    # Fetch allocations
    cursor.execute("""
        SELECT sa.id, sa.staff_id, sa.subject_id, sub.subject_code, sub.subject_name,
               sa.section_id, sec.section_name, d.dept_code, y.year_name
        FROM staff_assignments sa
        JOIN subjects sub ON sa.subject_id = sub.id
        JOIN sections sec ON sa.section_id = sec.id
        JOIN departments d ON sec.department_id = d.id
        JOIN years y ON sec.year_id = y.id;
    """)
    assignments = [dict(r) for r in cursor.fetchall()]
    conn.close()

    # Group assignments by staff
    assignment_map = {}
    for a in assignments:
        sid = a["staff_id"]
        if sid not in assignment_map:
            assignment_map[sid] = []
        assignment_map[sid].append(a)

    for s in staff:
        s["assignments"] = assignment_map.get(s["id"], [])

    return jsonify({"staff": staff}), 200

@admin_bp.route("/staff", methods=["POST"])
def create_staff():
    """Create a new staff account."""
    data = request.get_json() or {}
    username = data.get("username", "").strip()
    password = data.get("password", "").strip()
    full_name = data.get("full_name", "").strip()
    email = data.get("email", "").strip()
    dept_id = data.get("department_id")

    if not username or not password or not full_name or not email or not dept_id:
        return jsonify({"error": "All fields (username, password, full_name, email, department_id) are required"}), 400

    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        pw_hash = hash_password(password)
        cursor.execute("""
            INSERT INTO users (username, password_hash, full_name, email, role, department_id)
            VALUES (?, ?, ?, ?, 'staff', ?);
        """, (username, pw_hash, full_name, email, dept_id))
        conn.commit()
        new_id = cursor.lastrowid
        conn.close()
        return jsonify({"message": "Staff user created successfully", "id": new_id}), 201
    except Exception as e:
        conn.close()
        return jsonify({"error": f"Failed to register staff: {str(e)}"}), 400

@admin_bp.route("/staff-assignments", methods=["POST"])
def assign_staff():
    """Assign staff member to a subject and section."""
    data = request.get_json() or {}
    staff_id = data.get("staff_id")
    subject_id = data.get("subject_id")
    section_id = data.get("section_id")
    acad_year_id = data.get("academic_year_id")

    if not all([staff_id, subject_id, section_id, acad_year_id]):
        return jsonify({"error": "staff_id, subject_id, section_id, and academic_year_id are required"}), 400

    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute("""
            INSERT INTO staff_assignments (staff_id, subject_id, section_id, academic_year_id)
            VALUES (?, ?, ?, ?);
        """, (staff_id, subject_id, section_id, acad_year_id))
        conn.commit()
        new_id = cursor.lastrowid
        conn.close()
        return jsonify({"message": "Staff assigned successfully", "id": new_id}), 201
    except Exception as e:
        conn.close()
        return jsonify({"error": f"Assignment conflict or error: {str(e)}"}), 400

@admin_bp.route("/students", methods=["GET"])
def get_students():
    """List all students across sections with linked parent contact details."""
    section_id = request.args.get("section_id")
    conn = get_db_connection()
    cursor = conn.cursor()

    query = """
        SELECT s.id, s.register_number, s.roll_number, s.first_name, s.last_name,
               sec.section_name, d.dept_code, y.year_name,
               p.father_name, p.mother_name, p.primary_phone, p.whatsapp_number, p.email as parent_email
        FROM students s
        JOIN sections sec ON s.section_id = sec.id
        JOIN departments d ON sec.department_id = d.id
        JOIN years y ON sec.year_id = y.id
        JOIN parents p ON s.parent_id = p.id
    """
    params = []
    if section_id:
        query += " WHERE s.section_id = ?"
        params.append(section_id)

    query += " ORDER BY s.register_number ASC;"
    cursor.execute(query, params)
    students = [dict(r) for r in cursor.fetchall()]
    conn.close()

    return jsonify({"students": students}), 200

@admin_bp.route("/audit-logs", methods=["GET"])
def get_audit_logs():
    """Retrieve immutable audit trail of mark alterations."""
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("""
        SELECT al.id, al.previous_marks, al.new_marks, al.assessment_type, al.change_reason, al.created_at,
               s.register_number, s.first_name || ' ' || s.last_name as student_name,
               sub.subject_code, sub.subject_name,
               u.full_name as changed_by_name, u.role as changed_by_role
        FROM mark_audit_logs al
        JOIN students s ON al.student_id = s.id
        JOIN subjects sub ON al.subject_id = sub.id
        JOIN users u ON al.changed_by = u.id
        ORDER BY al.created_at DESC
        LIMIT 100;
    """)
    logs = [dict(r) for r in cursor.fetchall()]
    conn.close()

    return jsonify({"audit_logs": logs}), 200

@admin_bp.route("/notification-logs", methods=["GET"])
def get_notification_logs():
    """Retrieve parent dispatch logs (SMS and WhatsApp)."""
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("""
        SELECT nl.id, nl.recipient_phone, nl.channel, nl.notification_type, nl.message_body,
               nl.status, nl.created_at, u.full_name as triggered_by_name,
               s.register_number, s.first_name || ' ' || s.last_name as student_name
        FROM notification_logs nl
        LEFT JOIN students s ON nl.student_id = s.id
        JOIN users u ON nl.triggered_by = u.id
        ORDER BY nl.created_at DESC
        LIMIT 100;
    """)
    logs = [dict(r) for r in cursor.fetchall()]
    conn.close()

    return jsonify({"notification_logs": logs}), 200

File: server/src/routes/admin.routes.ts
import { Router } from 'express';
import { AdminController } from '../controllers/admin.controller';
import { StaffManagementController } from '../controllers/staffManagement.controller';
import { SubjectManagementController } from '../controllers/subjectManagement.controller';
import { AssessmentManagementController } from '../controllers/assessmentManagement.controller';
import { MarksUploadController } from '../controllers/marksUpload.controller';
import { AcademicStructureController } from '../controllers/academicStructure.controller';
import { StudentController } from '../controllers/student.controller';
import { StudentImportController } from '../controllers/studentImport.controller';
import { PerformanceController } from '../controllers/performance.controller';
import { AdminAnalyticsController } from '../controllers/adminAnalytics.controller';
import { NoticeController } from '../controllers/notice.controller';
import { NotificationTemplateController } from '../controllers/notificationTemplate.controller';
import { ReportController } from '../controllers/report.controller';
import multer from 'multer';
import { authenticateUser, requireAdmin } from '../middleware/auth.middleware';
import { prisma } from '../db';
import { config } from '../config';

const router = Router();

// Upload middleware with configurable file size limit
const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: (config.maxUploadSizeMb || 25) * 1024 * 1024 }
});

// Protect ALL routes in this router with authenticateUser AND requireAdmin
// Staff attempting to access any of these will receive 403 Forbidden
router.use(authenticateUser);
router.use(requireAdmin);

// Admin Dashboard Summary (Unified Institutional Command Center)
router.get('/dashboard-summary', AdminController.getDashboardSummary);
router.get('/dashboard', AdminController.getDashboardSummary);

// Admin Profile & Credentials Management
router.get('/profile', AdminController.getAdminProfile);
router.put('/profile/credentials', AdminController.updateAdminCredentials);

// Institutional Overview & Analytics (Strictly Admin-Only)
router.get('/analytics/overall', AdminAnalyticsController.getOverallAnalytics);
router.get('/analytics/departments', AdminAnalyticsController.getDepartmentAnalytics);
router.get('/analytics/department', AdminAnalyticsController.getDepartmentAnalytics);
router.get('/analytics/years', AdminAnalyticsController.getYearWiseAnalytics);
router.get('/analytics/year', AdminAnalyticsController.getYearWiseAnalytics);
router.get('/analytics/sections', AdminAnalyticsController.getSectionAnalytics);
router.get('/analytics/section', AdminAnalyticsController.getSectionAnalytics);
router.get('/analytics/drilldown', AdminAnalyticsController.getDrilldown);
router.get('/analytics/performance', PerformanceController.getPerformanceOverview);
router.get('/analytics/performance/overview', PerformanceController.getPerformanceOverview);
router.get('/analytics/performance/student/:studentId', PerformanceController.getStudentPerformance);
router.get('/analytics/performance/section/:sectionId', PerformanceController.getSectionPerformance);
router.post('/analytics/performance/snapshots/generate', PerformanceController.generateSnapshots);

// ----------------------------------------------------
// INSTITUTIONAL REPORTS (PDF, Excel, CSV, PPTX, JSON Preview)
// ----------------------------------------------------
router.get('/reports/types', ReportController.getReportCatalog);
router.get('/reports/generate', ReportController.generateAdminReport);
router.get('/reports/:reportType', ReportController.generateAdminReport);

// ----------------------------------------------------
// COMPLETE AUDIT LOGGING & MARK CHANGE HISTORY (Admin-Only, Immutable)
// ----------------------------------------------------
router.get('/audit-logs/mark-changes', AdminController.getMarkChangeLogs);
router.get('/audit-logs', AdminController.getAuditLogs);
router.delete('/audit-logs/mark-changes/:id', AdminController.preventAuditDeletion);
router.delete('/audit-logs/mark-changes', AdminController.preventAuditDeletion);
router.delete('/audit-logs/:id', AdminController.preventAuditDeletion);
router.delete('/audit-logs', AdminController.preventAuditDeletion);

// ----------------------------------------------------
// COLLEGE NOTICES & NEWS (Strictly Admin-Only)
// ----------------------------------------------------
router.get('/notices/recipients-preview', NoticeController.getRecipientsPreview);
router.get('/notices', NoticeController.getNotices);
router.post('/notices', NoticeController.createNotice);
router.get('/notices/:id', NoticeController.getNoticeById);
router.put('/notices/:id', NoticeController.updateNotice);
router.delete('/notices/:id', NoticeController.deleteNotice);
router.post('/notices/:id/publish', NoticeController.publishNotice);
router.post('/notices/:id/cancel', NoticeController.cancelNotice);
router.get('/notices/:id/notification-stats', NoticeController.getNotificationStats);

// ----------------------------------------------------
// NOTIFICATION TEMPLATES & PARENT MESSAGE PREVIEW & QUEUE
// ----------------------------------------------------
router.post('/notifications/preview-template', NotificationTemplateController.previewTemplate);
router.get('/notifications/preview-student-performance/:studentId', NotificationTemplateController.previewStudentPerformance);
router.get('/notifications/templates/samples', NotificationTemplateController.getSampleTemplates);
router.post('/notifications/send-performance', NotificationTemplateController.sendPerformanceNotification);
router.post('/notifications/send-section-performance', NotificationTemplateController.sendSectionPerformanceNotification);
router.get('/notifications/stats', NotificationTemplateController.getOverallQueueStats);
router.get('/notifications/history', NotificationTemplateController.getNotificationHistory);
router.post('/notifications/:id/retry', NotificationTemplateController.retryNotification);
router.post('/notifications/retry-failed', NotificationTemplateController.bulkRetryFailedNotifications);

// Marks Upload & Management (Admin Full Clearance)
router.get('/marks/template', MarksUploadController.downloadTemplate);
router.post('/marks/upload-preview', upload.single('file'), MarksUploadController.previewMarksUpload);
router.post('/marks/upload-confirm', MarksUploadController.confirmMarksUpload);
router.post('/marks/batch', MarksUploadController.batchManualEntry);

// ----------------------------------------------------
// SUBJECT MANAGEMENT (Admin-Only)
// ----------------------------------------------------
router.get('/subjects', SubjectManagementController.getSubjects);
router.get('/subjects/:id', SubjectManagementController.getSubjectById);
router.post('/subjects', SubjectManagementController.createSubject);
router.put('/subjects/:id', SubjectManagementController.updateSubject);
router.patch('/subjects/:id/status', SubjectManagementController.toggleSubjectStatus);
router.delete('/subjects/:id', SubjectManagementController.deleteSubject);

// ----------------------------------------------------
// ASSESSMENT MANAGEMENT (Admin-Only Dynamic Assessments)
// ----------------------------------------------------
router.get('/assessments', AssessmentManagementController.getAssessments);
router.get('/assessments/:id', AssessmentManagementController.getAssessmentById);
router.post('/assessments', AssessmentManagementController.createAssessment);
router.put('/assessments/:id', AssessmentManagementController.updateAssessment);
router.patch('/assessments/:id/status', AssessmentManagementController.toggleAssessmentStatus);
router.delete('/assessments/:id', AssessmentManagementController.deleteAssessment);

// ----------------------------------------------------
// FACULTY & STAFF MANAGEMENT (Admin-Only)
// ----------------------------------------------------
router.get('/staff', StaffManagementController.getStaffList);
router.post('/staff', StaffManagementController.createStaff);

// Teacher Assignments (Staff + Academic Year + Department + Year + Section + Subject)
// (Must precede dynamic /staff/:id routes)
router.get('/staff/assignments', StaffManagementController.getAllAssignments);
router.post('/staff/assignments', StaffManagementController.assignStaff);
router.put('/staff/assignments/:assignmentId', StaffManagementController.updateStaffAssignment);
router.patch('/staff/assignments/:assignmentId/change-course', StaffManagementController.updateStaffAssignment);
router.delete('/staff/assignments/:assignmentId', StaffManagementController.deleteStaffAssignment);

// Dynamic Staff Parametric Routes
router.get('/staff/:id', StaffManagementController.getStaffById);
router.get('/staff/:id/credentials', StaffManagementController.getStaffCredentialsSummary);
router.put('/staff/:id', StaffManagementController.updateStaff);
router.patch('/staff/:id/status', StaffManagementController.toggleStaffStatus);
router.patch('/staff/:id/discontinue', StaffManagementController.discontinueStaff);
router.post('/staff/:id/reissue-credentials', StaffManagementController.reissueStaffCredentials);
router.post('/staff/:id/reset-password', StaffManagementController.resetStaffPassword);
router.post('/staff/:id/authorize-password', StaffManagementController.authorizeStaffPassword);
router.delete('/staff/:id', StaffManagementController.deleteStaff);
router.get('/staff/:id/assignments', StaffManagementController.getStaffAssignments);
router.post('/staff/:id/assignments', StaffManagementController.assignStaff);


// ----------------------------------------------------
// ACADEMIC STRUCTURE MANAGEMENT (Admin-Only Hierarchy)
// ----------------------------------------------------

// Cascade Overview & Dependent Dropdown Helper
router.get('/academic-structure/overview', AcademicStructureController.getOverview);

// 1. Academic Years
router.get('/academic-years', AcademicStructureController.getAcademicYears);
router.post('/academic-years', AcademicStructureController.createAcademicYear);
router.put('/academic-years/:id', AcademicStructureController.updateAcademicYear);
router.patch('/academic-years/:id/status', AcademicStructureController.toggleAcademicYearStatus);
router.delete('/academic-years/:id', AcademicStructureController.deleteAcademicYear);

// 2. Departments
router.get('/departments', AcademicStructureController.getDepartments);
router.post('/departments', AcademicStructureController.createDepartment);
router.put('/departments/:id', AcademicStructureController.updateDepartment);
router.patch('/departments/:id/status', AcademicStructureController.toggleDepartmentStatus);
router.delete('/departments/:id', AcademicStructureController.deleteDepartment);

// 3. Years
router.get('/years', AcademicStructureController.getYears);
router.post('/years', AcademicStructureController.createYear);
router.put('/years/:id', AcademicStructureController.updateYear);
router.patch('/years/:id/status', AcademicStructureController.toggleYearStatus);
router.delete('/years/:id', AcademicStructureController.deleteYear);

// 4. Sections (Filtered by AcademicYear, Department, Year)
router.get('/sections', AcademicStructureController.getSections);
router.post('/sections', AcademicStructureController.createSection);
router.put('/sections/:id', AcademicStructureController.updateSection);
router.patch('/sections/:id/status', AcademicStructureController.toggleSectionStatus);
router.delete('/sections/:id', AcademicStructureController.deleteSection);

// ----------------------------------------------------
// STUDENT & PARENT MANAGEMENT (Admin-Only)
// ----------------------------------------------------

// Bulk Import Routes (SheetJS / Excel & CSV)
router.get('/students/import/template', StudentImportController.downloadTemplate);
router.post('/students/import/preview', upload.single('file'), StudentImportController.previewImport);
router.post('/students/import/confirm', StudentImportController.confirmImport);
router.get('/students/import/history', StudentImportController.getImportHistory);
router.get('/students/import/history/:id', StudentImportController.getImportHistoryDetails);

// Individual Student Management
router.get('/students', StudentController.getStudents);
router.get('/students/:id', StudentController.getStudentById);
router.post('/students', StudentController.createStudent);
router.put('/students/:id', StudentController.updateStudent);
router.patch('/students/:id/status', StudentController.toggleStudentStatus);
router.delete('/students/:id', StudentController.deleteStudent);

// ----------------------------------------------------
// PARENT DIRECTORY MANAGEMENT (Admin-Only)
// ----------------------------------------------------
router.get('/parents', async (req, res) => {
  try {
    const parents = await prisma.student.findMany({
      where: { status: 'ACTIVE' },
      select: {
        id: true,
        parentName: true,
        parentMobile: true,
        name: true,
        registerNumber: true,
        section: {
          select: {
            name: true,
            department: { select: { code: true } },
            year: { select: { name: true } }
          }
        }
      },
      orderBy: { parentName: 'asc' }
    });
    return res.status(200).json({ parents, total: parents.length });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to retrieve parents directory' });
  }
});

// ----------------------------------------------------
// SYSTEM SETTINGS MANAGEMENT (Admin-Only)
// ----------------------------------------------------
router.get('/settings', async (req, res) => {
  try {
    const academicYears = await prisma.academicYear.findMany({ orderBy: { yearName: 'desc' } });
    const currentYear = academicYears.find(y => y.isCurrent) || academicYears[0];
    return res.status(200).json({
      settings: {
        institutionName: 'VSB Engineering College',
        portalVersion: '2.0.0',
        currentAcademicYear: currentYear?.yearName || '2025-2026',
        currentAcademicYearId: currentYear?.id,
        gradingPolicy: {
          passPercentage: 50,
          internalMaxMarks: 100
        },
        auditLogging: {
          enabled: true,
          immutableLogs: true
        },
        notifications: {
          mockMode: process.env.MOCK_NOTIFICATIONS === 'true' || !process.env.WHATSAPP_ACCESS_TOKEN,
          supportedChannels: ['SMS', 'WHATSAPP']
        }
      }
    });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to retrieve system settings' });
  }
});

router.put('/settings', async (req, res) => {
  try {
    const { passPercentage, internalMaxMarks } = req.body || {};
    return res.status(200).json({
      message: 'System settings updated successfully',
      settings: {
        institutionName: 'VSB Engineering College',
        portalVersion: '2.0.0',
        gradingPolicy: {
          passPercentage: passPercentage || 50,
          internalMaxMarks: internalMaxMarks || 100
        }
      }
    });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to update system settings' });
  }
});

export default router;


